lib/source_file.h

```diff
@@ -205,6 +216,8 @@ class SourceInfo final {
     const InputSources *sources = nullptr;
     SourcePosition start = SourcePosition();
     SourcePosition end = SourcePosition();
+    mutable std::vector<Util::Comment *> before;
+    mutable std::vector<Util::Comment *> after;
```

Comment:
Another point -- this make the `SourceInfo`, which is ubiquitous in the compiler, quite a bit larger (by 6 pointers with the common impl of `vector`). It would be good to have some benchmarks to see this does not unduly show down the compiler (not `p4fmt` backend) or increase its memory consumption too much.

Reply:
If we'd go this way, I'd strongly suggest using small vectors that could hold up to 1 entry inline. But overall I concur, that enlarging every IR node by this amount should be carefully benchmarked. We do create lots of nodes. There is huge malloc traffic everywhere. And while another 48 bytes might look small, it is an overhead paid on every node. Even if there are zero comments.

Also, comments are quite rare and here the price is paid by *every* node. Maybe the better solution is to use the technique similar to `TrailingObjects` in LLVM / clang.

Reply:
Instead of storing those prefix and suffix comments as a part of `SourceInfo`, I was thinking how about storing that extra object in a hashmap. `Hashmap<node-id, extra-comment-object>`. This would avoid the unnecessary overhead to the compiler.
Would this be a good and straightforward solution for the time being?

Reply:
What is your plan to update this side map on IR change? Overall, side long-living maps should be discouraged as they could easily go stale

Reply:
Just want to add some more context. We were thinking about using this side map only in p4fmt. With this constraint, there probably isn't any IR change expected I guess.

@snapdgn has also looked into using `TrailingObjects`. I agree that would be a better solution. My suggestion is to go with a simple solution for now, to unblock further p4fmt experimentation, as long as the solution doesn't influence the rest of P4C. Then switching to the `TrailingObjects` technique could be left as a future optimization.

@asl What do you think?

Reply:
if this side map is local, then certainly we're fine.